### PR TITLE
Quick Search: Recent Search bugfix

### PR DESCRIPTION
**Pull request by OrionsGuardian**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

The Quick Search tool in the top right corner of the Constellation/Netbeans Window was not finding recent searches after relaunching the application.
It would not find any results due to formatting and character processing done within Netbeans.

The code has been modified to use substitute characters in place of those that Netbeans processes incorrectly.
The code will also recognise if an incoming search is actually re-running a recent search and use appropriate logic for those cases.

### Alternate Designs

N/A

### Why Should This Be In Core?

Clicking on a Recent Search entry will now correctly find results when available, rather than coming back with a No results found message.

### Benefits

Useful to re-run a search on a graph when first opening it for the day.

### Possible Drawbacks

N/A

### Verification Process

Open a graph.
Use the Quick Search tool to find a node, a transaction, a help page, and a Data Access plugin.
Close Constellation and launch it again.
Re-open the same graph.
Click the Quick Search area and select a recent search entry.
Ensure it returns the correct result.

### Applicable Issues

#186

### CODE REVIEW

**Review by Delphinus8821 — APPROVED** [review]

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessSearchProvider.java [hunk 1]
```diff
@@ -65,18 +80,24 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
 
         plugins.entrySet().forEach(entry -> unorderedPlugins.put(entry.getKey(), entry.getValue().getValue()));
 
+        final String comparisonText = "".equals(prevPluginName) ? text : prevPluginName;
```
Comment (antares1470) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessSearchProvider.java:83]: `prevPluginName.isEmpty()`?

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessSearchProvider.java [hunk 2]
```diff
@@ -65,18 +80,24 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
-                .filter(plugin -> StringUtils.containsIgnoreCase(plugin.getName(), text))
+                .filter(plugin -> StringUtils.containsIgnoreCase(plugin.getName(), comparisonText))
                 .map(DataAccessPlugin::getName)
                 .sorted((a, b) -> a.compareToIgnoreCase(b))
                 .collect(Collectors.toList());
 
         for (final String name : pluginNames) {
-            if (!response.addResult(new ShowDataAccessPluginTask(name), name)) {
+            final String displayName = QuickSearchUtilities.CIRCLED_D + "  " + QuickSearchUtilities.replaceBrackets(name);
+            if (!"".equals(prevPluginName) && StringUtils.containsIgnoreCase(name, prevPluginName)) {
```
Comment (antares1470) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/panes/DataAccessSearchProvider.java:96]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/EdgeQuickSearchProvider.java [hunk 3]
```diff
@@ -58,7 +59,28 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
+                    final int attrPos = prevSearch.lastIndexOf(FindResult.SEPARATOR);
+                    if (attrPos > -1) {
+                        prevSearch = prevSearch.substring(0, attrPos);
+                    }                    
+                } else {
+                    // This is a recent search for a different category, so we can end the Edge search here
+                    return;
+                }
+            }            
+            final String convertedSearch = !"".equals(prevSearch) ? prevSearch : searchRequest;
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/EdgeQuickSearchProvider.java:82]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/EdgeQuickSearchProvider.java [hunk 4]
```diff
@@ -72,12 +94,31 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
 
             final List<FindResult> results = plugin.getResults();
+            final List<FindResult> matchList = new ArrayList<>();
             for (final FindResult item : results) {
                 if (item != null) {
                     // We have a valid result, so report:
-                    response.addResult(new SelectContent(graph, item), item.toString());
+                    final String subscriptId = QuickSearchUtilities.buildSubscriptFromID(Integer.toString(item.getID()));
+                    final String displayText = QuickSearchUtilities.CIRCLED_E + "  " + QuickSearchUtilities.replaceBrackets(item.toString()) + "   " + QuickSearchUtilities.LH_SUB_BRACKET + subscriptId + QuickSearchUtilities.RH_SUB_BRACKET;
+                    if ("".equals(prevSearch)) {
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/EdgeQuickSearchProvider.java:103]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/LinkQuickSearchProvider.java [hunk 5]
```diff
@@ -58,7 +59,28 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
+                    final int attrPos = prevSearch.lastIndexOf(FindResult.SEPARATOR);
+                    if (attrPos > -1) {
+                        prevSearch = prevSearch.substring(0, attrPos);
+                    }                    
+                } else {
+                    // This is a recent search for a different category, so we can end the Link search here
+                    return;
+                }
+            }            
+            final String convertedSearch = !"".equals(prevSearch) ? prevSearch : searchRequest;
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/LinkQuickSearchProvider.java:82]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/LinkQuickSearchProvider.java [hunk 6]
```diff
@@ -72,12 +94,31 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
 
             final List<FindResult> results = plugin.getResults();
+            final List<FindResult> matchList = new ArrayList<>();
             for (final FindResult item : results) {
                 if (item != null) {
                     // We have a valid result, so report:
-                    response.addResult(new SelectContent(graph, item), item.toString());
+                    final String subscriptId = QuickSearchUtilities.buildSubscriptFromID(Integer.toString(item.getID()));
+                    final String displayText = QuickSearchUtilities.CIRCLED_L + "  " + QuickSearchUtilities.replaceBrackets(item.toString()) + "   " + QuickSearchUtilities.LH_SUB_BRACKET + subscriptId + QuickSearchUtilities.RH_SUB_BRACKET;
+                    if ("".equals(prevSearch)) {
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/LinkQuickSearchProvider.java:103]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/NodeQuickSearchProvider.java [hunk 7]
```diff
@@ -58,7 +60,28 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
+                    final int attrPos = prevSearch.lastIndexOf(FindResult.SEPARATOR);
+                    if (attrPos > -1) {
+                        prevSearch = prevSearch.substring(0, attrPos);
+                    }                    
+                } else {
+                    // This is a recent search for a different category, so we can end the Node search here
+                    return;
+                }
+            }            
+            final String convertedSearch = !"".equals(prevSearch) ? prevSearch : searchRequest;
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/NodeQuickSearchProvider.java:83]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/NodeQuickSearchProvider.java [hunk 8]
```diff
@@ -72,15 +95,33 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
 
             final List<FindResult> results = plugin.getResults();
+            final List<FindResult> matchList = new ArrayList<>();
             for (final FindResult item : results) {
                 if (item != null) {
                     // We have a valid result, so report:
-                    response.addResult(new SelectContent(graph, item), item.toString());
+                    final String subscriptId = QuickSearchUtilities.buildSubscriptFromID(Integer.toString(item.getID()));
+                    final String displayText = QuickSearchUtilities.CIRCLED_N + "  " + QuickSearchUtilities.replaceBrackets(item.toString()) + "   " + QuickSearchUtilities.LH_SUB_BRACKET + subscriptId + QuickSearchUtilities.RH_SUB_BRACKET;
+                    if ("".equals(prevSearch)) {
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/NodeQuickSearchProvider.java:104]: same here

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java [hunk 9]
```diff
@@ -56,9 +58,30 @@ public class TransactionQuickSearchProvider implements SearchProvider {
...
+                    final int attrPos = prevSearch.lastIndexOf(FindResult.SEPARATOR);
+                    if (attrPos > -1) {
+                        prevSearch = prevSearch.substring(0, attrPos);
+                    }
+                } else {
+                    // This is a recent search for a different category, so we can end the Transaction search here
+                    return;
+                }
+            }
+            final String convertedSearch = !"".equals(prevSearch) ? prevSearch : searchRequest;
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java:83]: ditto

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java [hunk 10]
```diff
@@ -72,12 +95,31 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
 
             final List<FindResult> results = plugin.getResults();
+            final List<FindResult> matchList = new ArrayList<>();
             for (final FindResult item : results) {
                 if (item != null) {
                     // We have a valid result, so report:
-                    response.addResult(new SelectContent(graph, item), item.toString());
+                    final String subscriptId = QuickSearchUtilities.buildSubscriptFromID(Integer.toString(item.getID()));
+                    final String displayText = QuickSearchUtilities.CIRCLED_T + "  " + QuickSearchUtilities.replaceBrackets(item.toString()) + "   " + QuickSearchUtilities.LH_SUB_BRACKET + subscriptId + QuickSearchUtilities.RH_SUB_BRACKET;
+                    if ("".equals(prevSearch)) {
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java:104]: ditto

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java [hunk 11]
```diff
@@ -72,12 +95,31 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
+                        // can potentially match the search term on a different graph, store the matches
+                        matchList.add(item);
+                    }
                 }
             }
+            if (!matchList.isEmpty()) {
+                // should only return 1 result when using the recent search function
+                final FindResult result = matchList.get(0);
+                final String subscriptId = QuickSearchUtilities.buildSubscriptFromID(Integer.toString(result.getID()));
+                String displayText = QuickSearchUtilities.CIRCLED_T + "  " + QuickSearchUtilities.replaceBrackets(result.toString()) + "   " + QuickSearchUtilities.LH_SUB_BRACKET + subscriptId + QuickSearchUtilities.RH_SUB_BRACKET;
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find/quicksearch/TransactionQuickSearchProvider.java:120]: final

File: CoreHelp/src/au/gov/asd/tac/constellation/help/utilities/HelpSearchProvider.java [hunk 12]
```diff
@@ -57,19 +73,47 @@ public void evaluate(final SearchRequest request, final SearchResponse response)
...
+            String displayName = qs.replaceBrackets(fileName.replace("-", " "));
+            final int indexMD = displayName.lastIndexOf(".");
+            displayName = QuickSearchUtils.CIRCLED_H + "  " + displayName.substring(0, indexMD);
 
+            if (fileName.contains(text.toLowerCase()) && "".equals(prevFileName)) {
                 // Display the result and add a runnable for when it is clicked on 
                 if (!response.addResult(new HelpSearchProviderTask(fileName), displayName)) {
                     return;
                 }
+            } else if (!StringUtils.isBlank(prevFileName) && fileName.contains(prevFileName)){
```
Comment (antares1470) [inline CoreHelp/src/au/gov/asd/tac/constellation/help/utilities/HelpSearchProvider.java:87]: `StringUtils.isNotBlank`

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/qs/QuickSearchUtilities.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    
+    // Substitution characters for angled brackets and round brackets, used to address a Netbeans issue
+    private static final String LT_FULL = "\uff1c"; // <
+    private static final String GT_FULL = "\uff1e"; // >
+    private static final String OB_FULL = "\uff08"; // (
+    private static final String CB_FULL = "\uff09"; // )
+    
+    private QuickSearchUtilities(){
+        // Should not be instantiated. Should only use the static methods.
+    }
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/qs/QuickSearchUtilities.java:50]: perhaps an `IllegalStateException` should be thrown here?

**Review by antares1470 — CHANGES_REQUESTED** [review]
for the isEmpty feedback, you can alternatively use `StringUtils.isEmpty` or `StringUtils.isNotEmpty` if you need null protection

**Pull request merged**